stdlib/tests/collections/smt.rs

```diff
@@ -0,0 +1,190 @@
+    let mut store = MerkleStore::new();
+
+    // `a` has no conflicts on depth `16` (first tier)
+    let a = InsertedLeaf::new_seeded(&mut store, root, &mut seed, a_3, 16);
+
+    // `b` conflicts with `c` up to `16`; the target tier is `32`
+    let b = InsertedLeaf::new_seeded(&mut store, a.root, &mut seed, b_3, 32);
+    let c = InsertedLeaf::new_seeded(&mut store, b.root, &mut seed, c_3, 32);
+
+    // `d` conflicts with `e` up to `32`; the target tier is `48`
+    let d = InsertedLeaf::new_seeded(&mut store, c.root, &mut seed, d_3, 48);
+    let e = InsertedLeaf::new_seeded(&mut store, d.root, &mut seed, e_3, 48);
+
+    // assert all nodes are returned with the MASM implementation
+    assert_leaves(store, e.root, [a, b, c, d, e]);
+}
+
+// TEST HELPERS
+// ================================================================================================
+
+/// Asserts key/value opens to root, provided the advice map and store
+fn assert_smt_get_opens_correctly(
+    key: Word,
+    value: Word,
+    root: Word,
+    store: MerkleStore,
+    advice_map: &[([u8; 32], Vec<Felt>)],
+) {
+    let source = r#"
+        use.std::collections::smt
+
+        begin
+            exec.smt::get
+        end
+    "#;
+    let initial_stack = [
+        root[0].as_int(),
+        root[1].as_int(),
+        root[2].as_int(),
+        root[3].as_int(),
+        key[0].as_int(),
+        key[1].as_int(),
+        key[2].as_int(),
+        key[3].as_int(),
+    ];
+    let expected_output = [
+        value[3].as_int(),
+        value[2].as_int(),
+        value[1].as_int(),
+        value[0].as_int(),
+        root[3].as_int(),
+        root[2].as_int(),
+        root[1].as_int(),
+        root[0].as_int(),
+    ];
+    let advice_stack = [];
+    build_test!(source, &initial_stack, &advice_stack, store, advice_map.iter().cloned())
+        .expect_stack(&expected_output);
+}
+
+/// Build the advice map and execute the test for all leaves, asserting they are returned by the
+/// MASM library
+fn assert_leaves<I>(store: MerkleStore, root: Word, leaves: I)
+where
+    I: IntoIterator<Item = InsertedLeaf>,
+{
+    let leaves: Vec<_> = leaves.into_iter().collect();
+    let advice_map: Vec<_> = leaves
+        .iter()
+        .map(|leaf| (leaf.node.into_bytes(), leaf.value.to_vec()))
+        .collect();
+
+    // tests all leaves individually
+    for leaf in leaves {
+        assert_smt_get_opens_correctly(leaf.key, leaf.value, root, store.clone(), &advice_map);
+    }
+}
+
+/// A representation of the post-insertion state of a leaf
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct InsertedLeaf {
+    /// Key
+    pub key: Word,
+    /// Remaining key for the depth
+    pub remaining_key: Word,
+    /// Generated value
+    pub value: Word,
+    /// Index in which the node was inserted
+    pub index: NodeIndex,
+    /// Computed node value
+    pub node: Word,
+    /// New root after insertion
+    pub root: Word,
```

Comment:
Why do we need `remaining_key` here?

Reply:
`remaining_key` is quite helpful to debug as we have it immediately available for inspection.